fedbiomed/node/requests/_overlay.py

```diff
@@ -1,258 +1,542 @@
-            f"from file {default_key_file}"
-        ) from e
-    return private_key
 
+    def _create_channel(self, distant_node_id: str) -> None:
+        """Create channel for peering with a given distant peer node if it does not exist yet.
 
-_default_n2n_key = load_default_n2n_key()
+        Assumes it is called with already acquired _channel_keys_lock
 
-# chunk size must be less or equal (in bits) the smallest RSA key length used
-_SMALLEST_KEY = 2048
-_CHUNK_SIZE = int(_SMALLEST_KEY / 8)
+        If channel does not exist, create it and generate local key.
+        If channel already exists, do nothing.
 
+        Args:
+            distant_node_id: unique ID of the peer node
+        """
+        if distant_node_id not in self._channel_keys:
+            local_key = DHKey()
+            self._channel_manager.add(distant_node_id, local_key.export_private_key())
 
-def format_outgoing_overlay(message: Message) -> List[bytes]:
-    """Creates an overlay message payload from an inner message.
+            self._channel_keys[distant_node_id] = _N2nKeysEntry(local_key=local_key, ready_event=asyncio.Event())
 
-    Serialize, crypt, sign the inner message
+    async def get_keys(self, distant_node_id: str) -> Tuple[DHKey, DHKey]:
+        """Gets keys for peering with a given distant peer node on a channel
 
-    Args:
-        message: Inner message to send as overlay payload
+        Args:
+            distant_node_id: unique ID of the peer node
 
-    Returns:
-        Payload for overlay message
+        Returns:
+            A tuple consisting of the public key object for the channel and the private key
+                object for the channel (or None if not known yet)
+        """
+        async with self._channel_keys_lock:
+            self._create_channel(distant_node_id)
+            return self._channel_keys[distant_node_id].local_key, self._channel_keys[distant_node_id].distant_key
 
-    Raises:
-        FedbiomedNodeToNodeError: key is too short
-        FedbiomedNodeToNodeError: bad message type
-    """
-    # robustify from developer error (try to encapsulate a bad message type)
-    if not isinstance(message, InnerMessage):
-        raise FedbiomedNodeToNodeError(
-            f"{ErrorNumbers.FB324.value}: not an inner message"
-        )
+    async def set_distant_key(self, distant_node_id: str, public_key_pem: bytes) -> None:
+        """Sets distant (public) key of a channel for peering with a given distant peer node.
 
-    # consider encrypt-sign([message,node_id]) or other see
-    # https://theworld.com/~dtd/sign_encrypt/sign_encrypt7.html
+        Args:
+            distant_node_id: unique ID of the peer node
+            public_key_pem: public key in PEM format
+        """
+        async with self._channel_keys_lock:
+            # This case should not happen, but this ensures robustness/integrity
+            self._create_channel(distant_node_id)
```

Comment:
As you commented this case should not happen. If it happens then it means that node received a reply that it is not requested, so I would recommend to trust implementation and remove `create channel`.

Reply:
I would prefer to keep, so that `_ChannelKeys` does not depend on being properly used by higher layer to run without sending an exception (without calling `_create_channel`, a call to `set_distant_keys` with a non-existing `distant_node_id` raises an `Exception` on accessing a non-existing element in a table which is pretty obscure for the higher layers).

So for me this is better implementation to keep it as it makes code modular (class is self coherent) and robust.

But if you don't agree, I will remove it :-)

Reply:
I don't agree because this is a private class that is used by only other classes that are low layer, and the classes that uses this class should respect its usage and make sure that it is properly used.  I prefer to get an exception that shows that there is an implementation issue and helps to identify the problem.

Reply:
Sorry in the meantime I double-checked, and I think we need to handle this case.
This is also a matter of security.

We call `set_distant_key` when we received a `ChannelSetupReply`.
A malicious distant node colluding with researcher may forge a `ChannelSetupReply`
with a non existing `distant_node_id` to force an error in our node.
So if we remove it, we need to add additional code in the `_FinalChannelReply` for checking if the `node_id` exists.

I can do this implementation which is more explicit, but it adds a little more complexity ...

Reply:
I can remove the `_create_channel` from the `wait_ready_channel` as discussed in previous comment. This one only protects from a bug in the `OverlayChannel`

Reply:
I understand but isn't there any machinery to check request id? How requests and replies are associated? I believe  it is also serious security issue if you process replies without validating request id.

Reply:
A legitimate (real hospital) but malicious (wants to attack other hospital) holds the proper `request_id`, so checking it does not protect in that case.

But you are right that currently, for channel setup, the `request_id` is not checked in the `ChannelSetupReply`. As the message is signed by sender, the message cannot be forged (only legitimate distant node can create the message, so I can't see any advantage for him of forging the request_id). But a malicious MITM (eg: researcher) could use that to replay message.

So thank you for spotting it, I will implement check of `request_id` for `ChannelSetupReply`

Reply:
> A legitimate (real hospital) but malicious (wants to attack other hospital) holds the proper request_id, so checking it does not protect in that case.

I didn't understand then, because it is same for the node who wants a create error: `A malicious distant node colluding with researcher may forge a ChannelSetupReply with a non existing distant_node_id to force an error in our node.` Then, to avoid such error we allow malicious colluding node will be able to create an entry in the `ChannelKeys` while this is not requested by the node.  So, it seems to me that the normal behavior should be to check request id (or any other entry)  to verify if the node did send a channel request to that specific node if not than it should discard the reply, and do not call `set_distant_key`.  This avoid creating malicious node to create an error and also an entry in `ChannelKeys` as well as in db. 

However, I understand that you want to keep it there so it is okay for me to keep it.

Reply:
Thank you @srcansiz for your thorough review :-)

> However, I understand that you want to keep it there so it is okay for me to keep it.

I am now implementing:
- checking `request_id` in `ChannelSetupReply` matches a pending request
- checking `distant_node_id` in `ChannelSetupReply` matches an existing channel entry

>> A legitimate (real hospital) but malicious (wants to attack other hospital) holds the proper request_id, so checking it does not protect in that case.
>I didn't understand then, because it is same for the node who wants to create an error:

My point was: if we check that the `distant_node_id` matches an existing channel entry, what can a legitimate malicious distant node do by using a forged `request_id` ?
I could not find any interest to him. because this does not allow him anything more than with the real `request_id`
However, this may leave room for replay attacks (I'm not sure about that).
Anyway, you're right: better to check it. So I implemented it.